File: typescript/src/TopInterview150/Array-String/RemoveElement/RemoveElement.ts
export function removeElement(nums: number[], val: number): number {
  let lastIndex = nums.length - 1;
  for (let index = 0; index <= lastIndex; index++) {
    if (nums[index] === val) {
      while (nums[lastIndex] === val && lastIndex > index) {
        lastIndex--;
      }
      nums[index] = nums[lastIndex];
      nums[lastIndex] = val;
      lastIndex--;
    }
  }
  return lastIndex + 1;
}

File: scala/src/main/scala/jbaldassari/leetcode/topinterview150/arraystring/removeelement/RemoveElement.scala
package jbaldassari.leetcode.topinterview150.arraystring.removeelement

import scala.annotation.tailrec

/*
 * Strategy:
 * Work inward from both ends of the array,
 * skipping consecutive values that are the same
 */

object Solution {
  def removeElement(nums: Array[Int], valueToRemove: Int): Int = {
    @tailrec
    def removeElementTR(nums: Array[Int], valueToRemove: Int, startIndex: Int = 0, endIndex: Int = nums.length - 1): Int = {
      if (startIndex > endIndex) {
        return endIndex + 1
      }
      val startIndexIsValToRemove = nums(startIndex) == valueToRemove
      val endIndexIsValToRemove = nums(endIndex) == valueToRemove
      if (startIndexIsValToRemove && !endIndexIsValToRemove) {
        nums(startIndex) = nums(endIndex)
        nums(endIndex) = valueToRemove
      }
      removeElementTR(nums, valueToRemove,
        if (startIndexIsValToRemove) startIndex else startIndex + 1,
        if (endIndexIsValToRemove) endIndex - 1 else endIndex)
    }
    removeElementTR(nums, valueToRemove)
  }
}
